python/perf-kernels/multreduce_matmul_kernel.py

```diff
@@ -0,0 +1,56 @@
+import triton
+import triton.language as tl
+
+
+# Kernel that implements GEMM with explicit multiply-reduce instructions for small block sizes.
+# Based on **tune_gemm** `matmul_kernel` from commit `cf44637` (see `triton-mlir` branch).
+@triton.jit
+def multreduce_matmul_kernel(a_ptr, b_ptr, c_ptr, bias_ptr, M, N, K, stride_am, stride_ak, stride_bk, stride_bn,
+                             stride_cm, stride_cn, stride_bias, BLOCK_SIZE_M: tl.constexpr, BLOCK_SIZE_N: tl.constexpr,
+                             BLOCK_SIZE_K: tl.constexpr, BIAS: tl.constexpr, EVEN_K: tl.constexpr):
+    pid = tl.program_id(axis=0)
+    num_pid_n = tl.cdiv(N, BLOCK_SIZE_N)
+    pid_m = pid // num_pid_n
+    pid_n = pid % num_pid_n
+    offs_k = tl.arange(0, BLOCK_SIZE_K)
+    offs_am = (pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M))
+    offs_bn = (pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N))
+    a_ptrs = a_ptr + offs_am[:, None] * stride_am + offs_k[None, :] * stride_ak
+    b_ptrs = b_ptr + offs_k[:, None] * stride_bk + offs_bn[None, :] * stride_bn
+    if BIAS:
+        bias_ptrs = bias_ptr + offs_am * stride_bias
+        bias = tl.load(bias_ptrs, mask=offs_am < M, other=0.0)
+    acc_dtype = tl.float32 if a_ptr.type.element_ty != tl.int8 else tl.int32
+    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=acc_dtype)
+    for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
+        if EVEN_K:
+            a = tl.load(a_ptrs)
+            b = tl.load(b_ptrs)
+        else:
+            a = tl.load(a_ptrs, mask=offs_k[None, :] < K - k * BLOCK_SIZE_K, other=0.0)
+            b = tl.load(b_ptrs, mask=offs_k[:, None] < K - k * BLOCK_SIZE_K, other=0.0)
+        # TODO: `min_dot_size` isn't always (16, 16, 16), it's in fact architecture
+        #       dependent and data type dependent. Please check this source file:
+        #       `${TRITON_ROOT}/third_party/amd/backend/compiler.py`
+        #       (look for `min_dot_size` function).
+        #       Q: How can we implement this `if` statement more precisely?
+        #       A: Maybe we can use an approach similar to `EVEN_K`, in which the kernel caller
+        #          inspect all relevant information and provides a Boolean flag for the kernel.
+        if (BLOCK_SIZE_M < 16 or BLOCK_SIZE_N < 16) or BLOCK_SIZE_K < 16:
+            # Explicit multiply-reduce for small block sizes.
+            a = tl.reshape(a, (BLOCK_SIZE_M, BLOCK_SIZE_K, 1)).to(acc_dtype)
+            b = tl.reshape(b, (1, BLOCK_SIZE_K, BLOCK_SIZE_N)).to(acc_dtype)
+            accumulator += tl.sum(a * b, axis=1)
+        else:
+            # Triton dot product for other block sizes.
+            accumulator += tl.dot(a, b)
```

Comment:
You can also remove this. We don't want to use dot here at all.

Reply:
Done by 2bba0d0b5.

Reply:
@vgokhale, even after all the work I still don't understand the purpose of this kernel we created. Could you please share your thoughts about this?

The facts are:
* `tl.dot` doesn't work for small block sizes.
* We have a kernel that computes the required dot product without relying on `tl.dot`, i.e. relying on explicit elementwise multiplication and sum reduction operations.
* In terms of performance, the proposed kernel is worse than using `tl.dot` with the minimum block size.

What are the advantages of the proposed kernel? If I were a Triton user on AMD hardware, I'd use `tl.dot` with the minimum block size instead of explicit multiply-reduce approach.

Reply:
Vinayak answered this question to me in a 1:1 meeting.

The proposed kernel will be a reference for further compiler developments and can also serve as a baseline for TorchInductor improvements. Furthermore, we can analyze its performance again in the future as the compiler gets better over time.